File: Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;

public abstract class SExp
{
    public override string ToString()
    {
        return "SExp";
    }
}

public class Atom : SExp
{
    public string Value { get; }

    public Atom(string value)
    {
        Value = value;
    }

    public override string ToString()
    {
        return Value;
    }
}

public class SList : SExp
{
    public List<SExp> Children { get; }

    public SList(List<SExp> children)
    {
        Children = children;
    }

    public override string ToString()
    {
        return "(" + string.Join("", Children.Select(child => child.ToString())) + ")";
    }
}

public class Parser
{
    public static SExp Parse(string input)
    {
        var tokens = Tokenize(input);
        return ParseTokens(tokens);
    }

    private static List<string> Tokenize(string input)
    {
        return input.Replace("(", " ( ").Replace(")", " ) ").Split(" ", StringSplitOptions.RemoveEmptyEntries).ToList();
    }

    private static SExp ParseTokens(List<string> tokens)
    {
        if (tokens.Count == 0)
        {
            throw new ArgumentException("Unexpected end of input");
        }

        var token = tokens[0];
        tokens.RemoveAt(0);
        if (token == "(")
        {
            var children = new List<SExp>();
            while (tokens.Count > 0 && tokens[0] != ")")
            {
                children.Add(ParseTokens(tokens));
            }
            if (tokens.Count == 0)
            {
                throw new ArgumentException("Unmatched (");
            }
            tokens.RemoveAt(0); // Remove ")"
            return new SList(children);
        }
        else if (token == ")")
        {
            throw new ArgumentException("Unmatched )");
        }
        else
        {
            return new Atom(token);
        }
    }

    public static void Main(string[] args)
    {
        string input = "(+ 1 (* 2 3))";
        SExp result = Parse(input);
        Console.WriteLine(result);
    }
}

File: parser.py
class SExp:
    def __str__(self):
        return "SExp"

class Atom(SExp):
    def __init__(self, value):
        self.value = value
    
    def __str__(self):
        return str(self.value)

class SList(SExp):
    def __init__(self, children):
        self.children = children

    def __str__(self):
        return "(" + "".join(str(child) for child in self.children) + ")"

def tokenize(input_string):
    input_string = input_string.replace("(", " ( ").replace(")", " ) ")
    return [token for token in input_string.split() if token]

def parse(input_string):
    tokens = tokenize(input_string)
    return parse_tokens(tokens)

def parse_tokens(tokens):
    if not tokens:
        raise ValueError("Unexpected end of input")
    
    token = tokens.pop(0)
    if token == "(":
        children = []
        while tokens and tokens[0] != ")":
            children.append(parse_tokens(tokens))
        if not tokens:
            raise ValueError("Unmatched (")
        tokens.pop(0) # Remove ")"
        return SList(children)
    elif token == ")":
        raise ValueError("Unmatched )")
    else:
        return Atom(token)

if __name__ == "__main__":
    input_string = "(+ 1 (* 2 3))"    
    result = parse(input_string)
    print(result)
